### PR TITLE
Vanilla Attention

### PR DESCRIPTION
**Pull request by dashesy**
Add vanilla attention as baseline and to be used by all archs.
The vanilla attention does not use "permute" specializations either, because they too fail on older archs.
I plan to add fmha from cutlass (and look into permute specializations) but this interim solution can be used for tests and as a baseline to compare latency.

### CODE REVIEW

**Comment by dashesy** [comment]
The lint CI is broken, probably flake updated

**Comment by antinucleon** [comment]
Linter is fixed now. @terrychenism could you review it?

File: python/aitemplate/frontend/nn/attention.py [hunk 1]
```diff
@@ -130,7 +183,33 @@ def __init__(
...
+                from aitemplate.compiler.base import _TorchConstantTensorData
+
+                mask = torch.triu(
+                    torch.ones(seq_len, seq_len, dtype=torch.bool), 1
+                ).cuda()
+                causal_mask_pt = torch.zeros(seq_len, seq_len).cuda().half()
+                causal_mask_pt.masked_fill_(mask, float("-inf"))
+                causal_mask_pt = causal_mask_pt.unsqueeze(0)
+
+                causal_mask = Tensor(
```
Comment (terrychenism) [inline python/aitemplate/frontend/nn/attention.py:199]: can we move this part outside frontend? user can assign the mask tensor if needed.

**Review by terrychenism — CHANGES_REQUESTED** [review]
I'd recommend to build a new class for vanilla attention. We will do a big refactor for fused attention, including variable seqlen attention and remove flash attention.

File: python/aitemplate/frontend/nn/attention.py (hunk 1 above)
Comment (dashesy) [reply python/aitemplate/frontend/nn/attention.py:199]: Makes sense, will create a new class and update the PR. 
The mask tensor constant is to implement "causal" so should be compile-time

Comment (terrychenism) [reply python/aitemplate/frontend/nn/attention.py:199]: Yes, mask tensor can be assigned before or after model compiling. It'd be good to keep ait frontend clean and avoid pytorch related code.

Comment (dashesy) [reply python/aitemplate/frontend/nn/attention.py]: Created a separate class and now the mask should be passed.

File: python/aitemplate/frontend/nn/vanilla_attention.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+from ...compiler import ops
+from .. import Tensor
+from .dropout import Dropout
+from .linear import Linear
+from .module import Module
+from .parameter import Parameter
+
+# pylint: disable=C0103
+
+USE_CUDA = detect_target().name() == "cuda"
```
Comment (terrychenism) [inline python/aitemplate/frontend/nn/vanilla_attention.py:31]: USE_CUDA can be removed.

File: python/aitemplate/frontend/nn/vanilla_attention.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        mask_seq: sequence mask, default: ``0``.
+    """
+
+    def __init__(
+        self,
+        dim,
+        seq_len,
+        seq_len_kv,
+        num_heads,
+        qkv_bias=False,
```
Comment (terrychenism) [inline python/aitemplate/frontend/nn/vanilla_attention.py:237]: do you think it's worth to split qkv_bias to q_bias/k_bias/v_bias?

**Review by terrychenism — APPROVED** [review]
lgtm. thanks!

File: python/aitemplate/frontend/nn/vanilla_attention.py (hunk 3 above)
Comment (terrychenism) [reply python/aitemplate/frontend/nn/vanilla_attention.py:233]: i know some speech models used proj with no bias for key. 
Ok, let's keep it simple for now.

Comment (dashesy) [reply python/aitemplate/frontend/nn/vanilla_attention.py:233]: I am not familiar with speech models, good to know!
